ENG-2178: Add download support

BartoszBlizniak opened the pull request:
This change should allow downloading the raw files that were uploaded to Cloudsmith using Terraform.

paddycarey commented on cloudsmith/data_source_package_download.go:
```diff
@@ -0,0 +1,150 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+// Hit the API with function PackagesListExecute with the following query: name and version (defaults to latest unless otherwise specified)
```
convention for function level comments in Go is to start with the function name e.g. `// getPackageURL does the thing ...`

paddycarey commented on README.md:
````diff
@@ -89,6 +89,32 @@ output "packages" {
 }
 ```
 
+Download the raw file for a package, the package version should equal to the package version visible in UI:
````
is this `raw` specific?

paddycarey commented on cloudsmith/data_source_package_download_test.go:
```diff
@@ -0,0 +1,196 @@
+					createHelloWorldFile(fileName)
+					uploadPackageToCloudsmith(apiKey, namespace, repoName, fileName, packageName, packageVersion, userName)
+				},
+				Config: testAccPackageDownloadData(namespace, repoName, packageName, packageVersion, destinationPath),
```
i think this is going to destroy your repository as you've removed it from the config here, that doesn't seem like what you want

paddycarey commented on cloudsmith/data_source_package_download.go:
```diff
@@ -0,0 +1,150 @@
+
+	queryString := query
+
+	if packageVersion != "" {
+		queryWithVersion := fmt.Sprintf("%s version:%s", query, packageVersion)
+		queryString = queryWithVersion
+	}
```
why treat version specially if all we're doing is appending it to the query string anyway? would it be simpler to just allow the user to specify the full query they need? could be much more flexible that way

paddycarey replied:
it does feel unnecessary to me tbh, especially in the context of a TF provider, the query field alone leaves things fully flexible and even lets users play with searches/filters in the UI to build the query they want

i expect a big usecase for this might also be pulling packages by hash or other params rather than version.. i can think of a bunch of cases where version alone probably wouldn't be enough e.g. pulling a debian package you probably want to specify a distro, a python package you might want to specify wheel or sdist, etc

it's easy to add version later in a backwards compatible way if we decide that's desirable, but i'm not sure the reverse is true

paddycarey commented on cloudsmith/data_source_package_download.go:
```diff
@@ -0,0 +1,150 @@
+	}
+
+	req = req.Query(queryString)
+	req = req.PageSize(1)
```
what if the list of returned packages extends beyond a single page?

paddycarey replied:
what was the resolution here?

paddycarey commented on cloudsmith/data_source_package_download.go:
```diff
@@ -0,0 +1,150 @@
+		return "", "", err
+	}
+
+	for _, pkg := range packagesList {
+		if pkg.GetName() == packageName {
+			return pkg.GetCdnUrl(), pkg.GetFilename(), nil
+		}
+	}
```
what if there's more than one match for the given package name and query?

paddycarey commented on cloudsmith/data_source_package_download.go:
```diff
@@ -0,0 +1,150 @@
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s/%s", namespace, repository, packageName))
```
would it make more sense to use the package's ID/slug_perm here?

paddycarey commented on cloudsmith/data_source_package_download.go:
```diff
@@ -0,0 +1,150 @@
+			"destination_path": {
+				Type:         schema.TypeString,
+				Description:  "The path to download the package to.",
+				Required:     true,
```
should we default this to a temp dir if not set?

paddycarey replied:
or just not download it if not set maybe?

paddycarey replied:
what was the resolution here?

paddycarey commented on cloudsmith/data_source_package_download_test.go:
```diff
@@ -0,0 +1,196 @@
+	sha256Checksum := fmt.Sprintf("%x", sha256.Sum256(fileData))
+
+	// Step 1: PUT (upload) the file
+	uploadURL := fmt.Sprintf("https://upload.cloudsmith.io/%s/%s/%s", namespace, repoName, fileName)
```
can we avoid this hardcoded URL here? otherwise the provider won't work against stg/dev/local envs

paddycarey commented on cloudsmith/data_source_package_download_test.go:
```diff
@@ -0,0 +1,196 @@
+	identifier := uploadResponse["identifier"]
+
+	// Step 2: POST the package details
+	createPackageURL := fmt.Sprintf("https://api-prd.cloudsmith.io/v1/packages/%s/%s/upload/raw/", namespace, repoName)
```
can we avoid this hardcoded URL here? otherwise the provider won't work against stg/dev/local envs

paddycarey commented on cloudsmith/data_source_package_download.go:
```diff
@@ -0,0 +1,150 @@
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"namespace": {
+				Type:         schema.TypeString,
+				Description:  "The namespace to which the package belongs.",
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
```
can we use organization rather than namesapce here? i know the provider is inconsistent, but so is the app itself, and we're slowing migrating away from `namespace` over time

BartoszBlizniak commented:
Hey @paddycarey - when you have time can you give it a second look?

I've added the base URL's as normal variables for now in order to check if the test will Pass. Once they're added to the secrets I can update the code. 

Also, added package sync wait.

I had to resort to the creation/removal of the repository via API as I kept getting this error when I have added creation of the repository through Terraform, I couldn't find a fix for this so this could be used as a workaround, unless you have any ideas as to why it's doing this:

```
Error in getPackageURL: 404 Not Found
    data_source_package_download_test.go:30: Step 2/2 error: Error running post-apply plan: exit status 1
        
        Error: 404 Not Found
        
          with data.cloudsmith_package_download.test,
          on terraform_plugin_test.tf line 3, in data "cloudsmith_package_download" "test":
           3: data "cloudsmith_package_download" "test" {
        
--- FAIL: TestAccPackageDownload_data (380.58s)
```

paddycarey commented on cloudsmith/data_source_package_download_test.go:
```diff
@@ -0,0 +1,324 @@
+					client := &http.Client{}
+					waitForPackageSync(client, apiKey, namespace, repoName, slugPerm, userName)
+
+					defer deleteRepository(apiKey, namespace, repoSlug, userName) // Delete repository after test
```
this is going to delete the repository immediately when this function finishes, not when the test finishes, is that what you want?

paddycarey commented:
> I had to resort to the creation/removal of the repository via API as I kept getting this error when I have added creation of the repository through Terraform, I couldn't find a fix for this so this could be used as a workaround, unless you have any ideas as to why it's doing this:

The ~300s timing for your test implies you're deleting the repository prior to attempting to use it. I'd check if something funny's going on there.

BartoszBlizniak commented:
> > I had to resort to the creation/removal of the repository via API as I kept getting this error when I have added creation of the repository through Terraform, I couldn't find a fix for this so this could be used as a workaround, unless you have any ideas as to why it's doing this:
> 
> The ~300s timing for your test implies you're deleting the repository prior to attempting to use it. I'd check if something funny's going on there.

So prior to adding the API calls to create/remove the repository, I spent 2 days trying to figure that out but couldn't really find a reason as to why this is happening. The overall code worked fine, it created the file, uploaded and downloaded correctly, but it would always fail just the test with a 404. I also added printout statements to each API call and all of them returned 200 during execution. 

Even during the execution of the code I noticed that the repository wasn't being deleted until the package was downloaded, so the code _worked_ but the test kept on failing, as soon as I comment out for creating the repository (with terraform) it passed the tests.

paddycarey commented:
What you probably want is one test step to create the repo, another to upload the file, and a third to add the data source to download it, then let terraform destroy the resources as normal. It feels like there's a lot of code here trying to work around a problem with the tests as written.

You definitely shouldn't need any custom code to create or destroy a repo, that's exactly what the provider itself is designed to do.

BartoszBlizniak commented:
@paddycarey - thanks, I've uploaded the updated code.

paddycarey commented on cloudsmith/data_source_package.go:
```diff
@@ -32,6 +39,53 @@ func dataSourcePackageRead(d *schema.ResourceData, m interface{}) error {
 
 	d.SetId(fmt.Sprintf("%s_%s_%s", namespace, repository, pkg.GetSlugPerm()))
+
+	if download {
+		err := downloadPackage(pkg.GetCdnUrl(), outputPath, pc.GetAPIKey())
+		if err != nil {
+			return err
+		}
+	} else {
+		d.Set("output_path", pkg.GetCdnUrl())
+	}
```
what happens here if the user doesn't set `output_path`?

paddycarey commented on cloudsmith/data_source_package.go:
```diff
@@ -156,7 +156,7 @@ func dataSourcePackage() *schema.Resource {
 				Description: "The path to save the downloaded package",
 				Optional:    true,
-				Default:     "",
+				Default:     "/tmp/",
```
what if the user's on windows or some other platform? can this be OS agnostic? maybe https://github.com/golang/go/blob/master/src/os/file.go#L392 ?

paddycarey commented on cloudsmith/data_source_package.go:
```diff
@@ -156,7 +164,7 @@ func dataSourcePackage() *schema.Resource {
 				Description: "The path to save the downloaded package",
 				Optional:    true,
-				Default:     "/tmp/",
+				Default:     "",
```
can you just call `os.TempDir()` here directly? save the if statement above?

paddycarey commented on cloudsmith/data_source_package.go:
```diff
@@ -0,0 +1,185 @@
+	namespace := requiredString(d, "namespace")
+	repository := requiredString(d, "repository")
+	identifier := requiredString(d, "identifier")
+	download := d.Get("download").(bool)
+	outputPath := d.Get("output_path").(string)
```
can you use `requiredBool` and `requiredString` for these? they do the same thing and would keep things consistent

paddycarey commented on cloudsmith/data_source_package.go:
```diff
@@ -0,0 +1,185 @@
+
+	d.SetId(fmt.Sprintf("%s_%s_%s", namespace, repository, pkg.GetSlugPerm()))
+
+	if download {
+		err := downloadPackage(pkg.GetCdnUrl(), outputPath, pc.GetAPIKey())
+		if err != nil {
+			return err
+		}
+	} else {
+		d.Set("output_path", pkg.GetCdnUrl())
+	}
```
with the branches here we end up with different types of path for `output_path`, one is going to be the folder in which the file is downloaded, the other is going to be a full URL including the package/file name

wondering what we should do here, do we need a `download_dir`schema variable or something? then we can set `output_path` to `computed` and then just always set that to the right _full_ path... what do you think?

paddycarey commented on cloudsmith/data_source_package.go:
```diff
@@ -0,0 +1,185 @@
+
+	req.Header.Add("Authorization", fmt.Sprintf("Token %s", apiKey))
+
+	client := &http.Client{}
```
if you can reuse the http client and transport that's already init'ed in provider_config.go then we'll get nice things like debug logging and timeouts etc. for free

paddycarey reviewed (commented):
looking good! just a few small bits and i think we're nearly good to go

paddycarey commented on cloudsmith/data_source_package.go:
```diff
@@ -0,0 +1,191 @@
+
+	d.SetId(fmt.Sprintf("%s_%s_%s", namespace, repository, pkg.GetSlugPerm()))
+
+	var computedOutputPath string
+
+	if download {
+		err := downloadPackage(pkg.GetCdnUrl(), outputPath, pc.GetAPIKey())
+		if err != nil {
+			return err
+		}
+
+		computedOutputPath = outputPath
+	} else {
+		computedOutputPath = pkg.GetCdnUrl()
+	}
+
+	d.Set("output_path", computedOutputPath)
```
overriding output_path like this when it's already set is going to cause issues with Terraform complaining about inconsistency, you can see this if you add a test that uses the value

you probably need two schema variables, one to set the output dir, and then use output path to output the full path including file (whether local or remote)
